src/reducer.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  FormikErrors,
+  FormikState,
+  FormikValues,
+  FormikTouched,
+  setIn,
+} from '.';
+
+/**
+ * We probably should not use a constant enum here, but Babel 7 isn't out yet
+ */
+export enum FormikConstants {
+  SUBMIT_FORM = 'SUBMIT_FORM',
+  VALIDATE_ATTEMPT = 'VALIDATE_ATTEMPT',
+  VALIDATE_FAILURE = 'VALIDATE_FAILURE',
+  VALIDATE_SUCCESS = 'VALIDATE_SUCCESS',
+  RESET_FORM = 'RESET_FORM',
+  SET_FORMIK_STATE = 'SET_FORMIK_STATE',
+  SET_FIELD_VALUE = 'SET_FIELD_VALUE',
+  SET_FIELD_ERROR = 'SET_FIELD_ERROR',
+  SET_FIELD_TOUCHED = 'SET_FIELD_TOUCHED',
+  SET_VALUES = 'SET_VALUES',
+  SET_TOUCHED = 'SET_TOUCHED',
+  SET_ERRORS = 'SET_ERRORS',
+  SET_STATUS = 'SET_STATUS',
+  SET_SUBMITTING = 'SET_SUBMITTING',
+}
+
+export type Actions<Values> =
+  | {
+      type: FormikConstants.SUBMIT_FORM;
+    }
+  | {
+      type: FormikConstants.VALIDATE_ATTEMPT;
+    }
+  | {
+      type: FormikConstants.VALIDATE_SUCCESS;
+    }
+  | {
+      type: FormikConstants.VALIDATE_FAILURE;
+      payload: FormikErrors<Values>;
+    }
+  | {
+      type: FormikConstants.RESET_FORM;
+      payload?: FormikValues;
+    }
+  | {
+      type: FormikConstants.SET_FORMIK_STATE;
+      payload: FormikState<Values>;
+    }
+  | {
+      type: FormikConstants.SET_SUBMITTING;
+      payload: boolean;
+    }
+  | {
+      type: FormikConstants.SET_FIELD_VALUE;
+      payload: { field: string; value: any };
+    }
+  | {
+      type: FormikConstants.SET_FIELD_ERROR;
+      payload: { field: string; value?: string };
+    }
+  | {
+      type: FormikConstants.SET_FIELD_TOUCHED;
+      payload: { field: string; value: boolean };
+    }
+  | {
+      type: FormikConstants.SET_VALUES;
+      payload: FormikValues;
+    }
+  | {
+      type: FormikConstants.SET_TOUCHED;
+      payload: FormikTouched<Values>;
+    }
+  | {
+      type: FormikConstants.SET_STATUS;
+      payload?: any;
+    }
+  | {
+      type: FormikConstants.SET_ERRORS;
+      payload: FormikErrors<Values>;
+    };
+
+export function reducer<Values>(
```

Comment:
...so this lays the ground work for porting Formik to other JS frameworks. Hmmm. ![bitmoji](https://render.bitstrips.com/v2/cpanel/7e3bc453-64fb-4a34-bcaa-481724e5d3d9-d1a1e672-c5f5-466f-bff9-82e07b058b8f-v1.png?transparent=1&palette=1&width=246)